pkg/autoscaler/metrics/collector.go

```diff
@@ -285,58 +290,73 @@ func newCollection(metric *av1alpha1.Metric, scraper StatsScraper, tickFactory f
 			metric.Spec.PanicWindow, config.BucketSize),
 		scraper: scraper,
 
-		stopCh: make(chan struct{}),
+		tickFactory: tickFactory,
 	}
 
-	key := types.NamespacedName{Namespace: metric.Namespace, Name: metric.Name}
-	logger = logger.Named("collector").With(zap.String(logkey.Key, key.String()))
+	if metric.Spec.ScrapeTarget != "" {
+		c.startScraping()
+	}
+
+	return c
+}
 
+// startScraping starts scraping metrics.
+// This is only run from inside a lock, it can't race with a stopScraping call.
+func (c *collection) startScraping() {
+	if c.stopCh != nil {
+		// If scraper is already running, do nothing.
+		return
+	}
+	c.stopCh = make(chan struct{})
 	c.grp.Add(1)
 	go func() {
 		defer c.grp.Done()
 
-		scrapeTicker := tickFactory(scrapeTickInterval)
+		scrapeTicker := c.tickFactory(scrapeTickInterval)
 		defer scrapeTicker.Stop()
 		for {
 			select {
 			case <-c.stopCh:
 				return
 			case <-scrapeTicker.C:
-				currentMetric := c.currentMetric()
-				if currentMetric.Spec.ScrapeTarget == "" {
-					// Don't scrape empty target service.
-					if c.updateLastError(nil) {
-						callback(key)
-					}
-					continue
-				}
-				stat, err := c.getScraper().Scrape(currentMetric.Spec.StableWindow)
-				if err != nil {
-					logger.Errorw("Failed to scrape metrics", zap.Error(err))
-				}
+				metric := c.currentMetric()
+				stat, err := c.getScraper().Scrape(metric.Spec.StableWindow)
 				if c.updateLastError(err) {
-					callback(key)
+					c.callback(types.NamespacedName{Namespace: metric.Namespace, Name: metric.Name})
 				}
 				if stat != emptyStat {
 					c.record(stat)
 				}
 			}
 		}
 	}()
-
-	return c
 }
 
-// close stops collecting metrics, stops the scraper.
-func (c *collection) close() {
+// stopScraping stops the scraper.
+// This is only run from inside a lock, it can't race with a startScraping call.
+func (c *collection) stopScraping() {
 	close(c.stopCh)
```

Comment:
do we care to check `stopCh` isn't nil?

Reply:
Yep, that's the error I was alluding to when holding this.

Reply:
Let us know when it's ready :-)